### ISSUE TITLE
Performance: non-smooth losses (Max*, SumAbsolute) are 20–100× slower than smooth losses on the same fit

### ISSUE THREAD

**Issue by jolars** [jolars/eunoia#45]
## Summary

On a 6-set ellipse fit, switching the loss from `NormalizedSumSquared` (smooth) to `SumAbsolute` or any of the `Max*` losses makes the fitter 20–100× slower for the **same input, same seed, same shape**.

This was found while profiling eulerr's test suite against eunoia 0.5.0: a single 6-set ellipse fit with `loss = MaxAbsolute` took ~16 s (with `tolerance = 1e-8`), accounting for ~30% of the entire eulerr R-package test suite runtime by itself. Two fits in `test-reproducibility.R::s[[9]]` (one with `MaxAbsolute`, one with `MaxSquared`) sum to ~32 s of the 53 s suite.

## Reproduction (eunoia 0.5.0, release with `lto = true, codegen-units = 1`, default tolerance)

```rust
use eunoia::{
    geometry::shapes::Ellipse, loss::LossType, DiagramSpecBuilder, Fitter, InputType,
};
use std::io::Write;
use std::time::Instant;

fn six_set_spec() -> eunoia::spec::DiagramSpec {
    DiagramSpecBuilder::new()
        .set("A", 4.0).set("B", 6.0).set("C", 3.0)
        .set("D", 2.0).set("E", 7.0).set("F", 3.0)
        .intersection(&["A","B"], 2.0).intersection(&["A","F"], 2.0)
        .intersection(&["B","C"], 2.0).intersection(&["B","D"], 1.0)
        .intersection(&["B","F"], 2.0).intersection(&["C","D"], 1.0)
        .intersection(&["D","E"], 1.0).intersection(&["E","F"], 1.0)
        .intersection(&["A","B","F"], 1.0)
        .intersection(&["B","C","D"], 1.0)
        .input_type(InputType::Exclusive)
        .build()
        .unwrap()
}

fn bench(label: &str, n: usize, loss: LossType) {
    let spec = six_set_spec();
    print!("  {:<32}", label);
    std::io::stdout().flush().unwrap();
    let _ = Fitter::<Ellipse>::new(&spec).loss_type(loss).seed(1).fit().unwrap();
    let t = Instant::now();
    for i in 0..n {
        let _ = Fitter::<Ellipse>::new(&spec).loss_type(loss).seed(i as u64 + 2).fit().unwrap();
    }
    let per = t.elapsed().as_secs_f64() * 1000.0 / n as f64;
    println!(": {:>2} fits, {:>9.1} ms/fit", n, per);
}

fn main() {
    bench("loss=SumSquared",                 3, LossType::SumSquared);
    bench("loss=NormalizedSumSquared",       3, LossType::NormalizedSumSquared);
    bench("loss=SumAbsolute",                3, LossType::sum_absolute());
    bench("loss=MaxSquared",                 3, LossType::MaxSquared);
    bench("loss=MaxAbsolute",                3, LossType::MaxAbsolute);
}
```

## Numbers (default tolerance = 1e-6)

```
  loss=SumSquared                 :  3 fits,      22.1 ms/fit
  loss=NormalizedSumSquared       :  3 fits,      19.7 ms/fit
  loss=SumAbsolute                :  3 fits,    1932.7 ms/fit
  loss=MaxSquared                 :  3 fits,     420.3 ms/fit
  loss=MaxAbsolute                :  3 fits,     918.1 ms/fit
```

| Loss                     | ms/fit   | vs `NormalizedSumSquared` |
|--------------------------|---------:|--------------------------:|
| `NormalizedSumSquared`   |    19.7  | 1× (baseline)             |
| `SumSquared`             |    22.1  | 1.1×                      |
| `MaxSquared`             |   420.3  | 21×                       |
| `MaxAbsolute`            |   918.1  | 47×                       |
| `SumAbsolute`            |  1932.7  | 98×                       |

In the actual eulerr test path (`tolerance = 1e-8`, the eunoia default L-BFGS-pool optimizer), `MaxAbsolute` and `MaxSquared` blow up further to ~16 s/fit each — presumably because the tighter tolerance forces many more iterations of a step direction that's already poorly informed by the loss landscape.

## Likely cause

`MaxSquared`, `MaxAbsolute`, `SumAbsolute`, `DiagError`, `SumAbsoluteRegionError` are all **non-smooth**:

- `Max*` losses have zero gradient w.r.t. all but a single "active" region per evaluation; central-difference gradients return mostly zeros and L-BFGS makes no useful progress until the active region happens to flip.
- `SumAbsolute` (and absolute region error) are continuous but non-differentiable at any zero crossing, so the gradient discontinuity confuses both finite-diff Hessian estimates and quasi-Newton curvature updates.

The smooth (`SumSquared`, `NormalizedSumSquared`) cases evaluate fast because every cost eval produces a smooth gradient → L-BFGS converges in a few dozen iterations.

## Suggested directions

1. **Detect non-smooth losses in the fitter and dispatch differently.** For `Max*` and absolute losses, prefer Nelder-Mead (no gradient required) over L-BFGS — likely a 10–100× speedup at no quality cost. Could be as simple as a `LossType::is_smooth() -> bool` and a `Fitter` branch.
2. **Smooth approximation.** Replace `MaxAbsolute` internally with a soft-max (`logsumexp` style) above some tolerance to keep gradients useful, anneal the smoothing factor as the optimizer converges.
3. **Document the cost.** At minimum, adding a doc comment to each non-smooth `LossType` variant warning users about the cost-with-L-BFGS would let downstream callers (eulerr) avoid them in hot paths.

## Context

eulerr migration: backend switch from Rcpp/C++ to extendr/eunoia. `test-reproducibility.R` exercises every (shape, loss, loss_aggregator) combination over 17 specs; the two slow combos (`s[[9]]` × `ellipse` × `(square|abs, max)`) drive ~30 s of the 53 s suite runtime. Most users of eulerr won't pick `loss_aggregator = "max"`, but the perf gap is large enough that even the smooth-loss path could plausibly be improved.
